"Feature" request: Update files from Notepad++ automatically

It would be good if the files that come from Notepad++ (such as `Notepad_plus_msgs.h` and `Scintilla.h`) were updated automatically when they change in Notepad++. Anyone starting a new plugin from this template (or from plugindemo) should be starting with the most up-to-date versions of those files.

I do not know GitHub well enough to know how to do this, or even if it can be done.

I’m thinking that if this can be done, it should be recommended to users to start a new plugin by downloading from the Master branch (using the big green **Code** button on the main page and selecting **Dowload ZIP**) rather than downloading a release zip, so it wouldn’t be necessary to create a new release whenever these files change. (Though if it were also possible to create a new release automatically, and have the tag match the Notepad++ release it supports, that would be excellent.)

[Related community topic](https://community.notepad-plus-plus.org/topic/26834/)

@pryrt 

> [W]hen I include just the first 7 files [@Coises](https://github.com/Coises) mentioned, it copies and builds okay. If I do the additional 4 or 6, after grabbing the new copies of the files, the build fails.

More specifically, the current versions of the `DockingDlgInterface` and `StaticDialog` modules have acquired more dependencies since the last template release:

    MSBuild version 17.13.19+0d9f5a35a for .NET Framework

      GoToLineDlg.cpp
    C:\Users\Public\git\plugintemplate\src\DockingFeature\DockingDlgInterface.h(25,10): error C1083: Cannot open include file: 'Common.h': No suc
    h file or directory [C:\Users\Public\git\plugintemplate\vs.proj\NppPluginTemplate.vcxproj]
      (compiling source file '../src/DockingFeature/GoToLineDlg.cpp')

      StaticDialog.cpp
    C:\Users\Public\git\plugintemplate\src\DockingFeature\StaticDialog.h(18,10): error C1083: Cannot open include file: 'dpiManagerV2.h': No such
     file or directory [C:\Users\Public\git\plugintemplate\vs.proj\NppPluginTemplate.vcxproj]
      (compiling source file '../src/DockingFeature/StaticDialog.cpp')

    ...

I tried to follow the dependency tree to find out how many transitive headers would be needed; I gave up once I got to TinyXML ...😖 

Fortunately the template project will build with a smaller subset of the upstream code base. Here are some development scripts for anyone who wants to play around.

<details>
<summary>scripts</summary>

```py
# ./file_list.py
"""
TEMPLATE_ROOT : Path
    Local directory root of the template source files.

UPSTREAM_ROOT : Path
    Local directory root the upstream Notepad++ files.

SOURCE_FILE_MAP : tuple[str, str]
    Maps each template source file to a corresponding upstream file.

    :NOTE: Be careful when editing the file list -- the order matters!
"""
from pathlib import Path

TEMPLATE_ROOT = Path('.git/').parent
UPSTREAM_ROOT = TEMPLATE_ROOT / 'npp_repo' 

SOURCE_FILE_MAP = zip(
    [str(TEMPLATE_ROOT / p) for p in [
        'src/Notepad_plus_msgs.h',
        'src/PluginInterface.h',
        # 'src/Parameters.cpp',
        # 'src/Parameters.h',
        # 'src/DockingFeature/Common.h',
        # 'src/DockingFeature/NppDarkMode.cpp',
        # 'src/DockingFeature/NppDarkMode.h',
        'src/DockingFeature/Docking.h',
        # 'src/DockingFeature/DockingDlgInterface.h',
        'src/DockingFeature/dockingResource.h',
        # 'src/DockingFeature/dpiManagerV2.cpp',
        # 'src/DockingFeature/dpiManagerV2.h',
        # 'src/DockingFeature/StaticDialog.cpp',
        # 'src/DockingFeature/StaticDialog.h',
        'src/DockingFeature/Window.h',
        # 'src/DockingFeature/DarkMode/DarkMode.cpp',
        # 'src/DockingFeature/DarkMode/DarkMode.h',
        # 'src/DockingFeature/DarkMode/IatHook.h',
        # 'src/DockingFeature/DarkMode/UAHMenuBar.h',
        'src/menuCmdID.h',
        'src/Sci_Position.h',
        'src/Scintilla.h',
    ]],

    [str(UPSTREAM_ROOT / p) for p in [
        'PowerEditor/src/MISC/PluginsManager/Notepad_plus_msgs.h',
        'PowerEditor/src/MISC/PluginsManager/PluginInterface.h',
        # 'PowerEditor/src/Parameters.cpp',
        # 'PowerEditor/src/Parameters.h',
        # 'PowerEditor/src/MISC/Common/Common.h',
        # 'PowerEditor/src/NppDarkMode.cpp',
        # 'PowerEditor/src/NppDarkMode.h',
        'PowerEditor/src/WinControls/DockingWnd/Docking.h',
        # 'PowerEditor/src/WinControls/DockingWnd/DockingDlgInterface.h',
        'PowerEditor/src/WinControls/DockingWnd/dockingResource.h',
        # 'PowerEditor/src/dpiManagerV2.cpp',
        # 'PowerEditor/src/dpiManagerV2.h',
        # 'PowerEditor/src/WinControls/StaticDialog/StaticDialog.cpp',
        # 'PowerEditor/src/WinControls/StaticDialog/StaticDialog.h',
        'PowerEditor/src/WinControls/Window.h',
        # 'PowerEditor/src/DarkMode/DarkMode.cpp',
        # 'PowerEditor/src/DarkMode/DarkMode.h',
        # 'PowerEditor/src/DarkMode/IatHook.h',
        # 'PowerEditor/src/DarkMode/UAHMenuBar.h',
        'PowerEditor/src/menuCmdID.h',
        'scintilla/include/Sci_Position.h',
        'scintilla/include/Scintilla.h',
    ]])
```

```py
# ./sync_files.py
import sys
import subprocess

from subprocess import CalledProcessError
# see the 'file_list.py' module, above
from file_list import SOURCE_FILE_MAP, UPSTREAM_ROOT

def sync():
    """
    Copy each upstream source file into the corresponding template file path.

    :return: ``0`` if successful, otherwise a non-zero exit code.
    """
    if sys.platform.lower().startswith('win'):
        cmd = ['xcopy', '/KVY']
    else:
        cmd = ['cp', '-aT']

    if UPSTREAM_ROOT.is_dir():
        for ours, theirs in SOURCE_FILE_MAP:
            err = None
            commandline = cmd + [theirs, ours]
            try:
                subprocess.run(commandline, check=True)
            except CalledProcessError as proc_err:
                err = proc_err
            except (FileNotFoundError, IOError, OSError) as misc_err:
                err = misc_err

            if err is not None:
                exit_code = err.returncode if err is CalledProcessError else -1
                print(err, file=sys.stderr)
                return exit_code
    return 0


if __name__ == '__main__':
    sys.exit(sync())

```
</details>

--- a/.github/workflows/CI_update_remote.yml
+++ b/.github/workflows/CI_update_remote.yml
@@ -47,13 +47,14 @@ jobs:
             "src\DockingFeature\Docking.h" = "npp_repo\PowerEditor\src\WinControls\DockingWnd\Docking.h",
             "src\DockingFeature\dockingResource.h" = "npp_repo\PowerEditor\src\WinControls\DockingWnd\dockingResource.h"
             ##
-            ### Unknown: the following _might_ be needed, as well
+            ### The following might be useful to plugin developers, but may require changes to make them compatible
+            ###     as such, they are listed for reference, but not included as part of the automatic update
+            ###     of the template
             ##
             #"src\DockingFeature\DockingDlgInterface.h" = "npp_repo\PowerEditor\src\WinControls\DockingWnd\DockingDlgInterface.h",
             #"src\DockingFeature\StaticDialog.h" = "npp_repo\PowerEditor\src\WinControls\StaticDialog\StaticDialog.h",
             #"src\DockingFeature\StaticDialog.cpp" = "npp_repo\PowerEditor\src\WinControls\StaticDialog\StaticDialog.cpp",
             #"src\DockingFeature\Window.h" = "npp_repo\PowerEditor\src\WinControls\Window.h",
-            ##
             #"src\DockingFeature\GoToLineDlg.cpp" = "npp_repo\PowerEditor\src\ScintillaComponent\GoToLineDlg.cpp",
             #"src\DockingFeature\GoToLineDlg.h" = "npp_repo\PowerEditor\src\ScintillaComponent\GoToLineDlg.h",
         }

--- a/.github/workflows/CI_update_remote.yml
+++ b/.github/workflows/CI_update_remote.yml
@@ -36,25 +36,38 @@ jobs:
         # start assuming no differences
         $anyDifference = $false
         #
-        # scintilla hashes: for plugintemplate (hps) and npp (hns)
+        # Define the local_path => upstream_path pairs
         #
-        $fps = "src\Scintilla.h"
-        $fns = "npp_repo\scintilla\include\Scintilla.h"
-        $hps = Get-FileHash $fps -Algorithm SHA256 | Select-Object -ExpandProperty Hash
-        $hns = Get-FileHash $fns -Algorithm SHA256 | Select-Object -ExpandProperty Hash
-        if ($hps -ne $hns) {
-            $anyDifference = $true
-            Copy-Item -path $fns -destination $fps -force
+        $pairs = @{
+            "src\Scintilla.h" = "npp_repo\scintilla\include\Scintilla.h",
+            "src\Sci_Position.h" = "npp_repo\scintilla\include\Sci_Position.h",
+            "src\Notepad_plus_msgs.h" = "npp_repo\PowerEditor\src\MISC\PluginsManager\Notepad_plus_msgs.h"
+            "src\PluginInterface.h" = "npp_repo\PowerEditor\src\MISC\PluginsManager\PluginInterface.h"
+            "src\menuCmdID.h" = "npp_repo\PowerEditor\src\menuCmdID.h",
+            "src\DockingFeature\Docking.h" = "npp_repo\PowerEditor\src\WinControls\DockingWnd\Docking.h",
+            "src\DockingFeature\dockingResource.h" = "npp_repo\PowerEditor\src\WinControls\DockingWnd\dockingResource.h"
+            ##
+            ### Unknown: the following _might_ be needed, as well
+            ##
+            #"src\DockingFeature\DockingDlgInterface.h" = "npp_repo\PowerEditor\src\WinControls\DockingWnd\DockingDlgInterface.h",
+            #"src\DockingFeature\StaticDialog.h" = "npp_repo\PowerEditor\src\WinControls\StaticDialog\StaticDialog.h",
+            #"src\DockingFeature\StaticDialog.cpp" = "npp_repo\PowerEditor\src\WinControls\StaticDialog\StaticDialog.cpp",
+            #"src\DockingFeature\Window.h" = "npp_repo\PowerEditor\src\WinControls\Window.h",
+            ##
+            #"src\DockingFeature\GoToLineDlg.cpp" = "npp_repo\PowerEditor\src\ScintillaComponent\GoToLineDlg.cpp",
+            #"src\DockingFeature\GoToLineDlg.h" = "npp_repo\PowerEditor\src\ScintillaComponent\GoToLineDlg.h",
         }
         #
-        # Notepad_plus_msgs.h hashes: for plugintemplate (hpn) and npp (hnn)
-        $fpn = "src\Notepad_plus_msgs.h"
-        $fnn = "npp_repo\PowerEditor\src\MISC\PluginsManager\Notepad_plus_msgs.h"
-        $hpn = Get-FileHash $fpn -Algorithm SHA256 | Select-Object -ExpandProperty Hash
-        $hnn = Get-FileHash $fnn -Algorithm SHA256 | Select-Object -ExpandProperty Hash
-        if ($hpn -ne $hnn) {
-            $anyDifference = $true
-            Copy-Item -path $fnn -destination $fpn -force
+        # loop over each pair, calculate hash of each, and copy any files that are different
+        #
+        ForEach ($local_path in $pairs.Keys) {
+            $upstream_path = $pairs[$local_path]
+            $local_hash = Get-FileHash $local_path -Algorithm SHA256 | Select-Object -ExpandProperty Hash
+            $upstream_hash = Get-FileHash $upstream_path -Algorithm SHA256 | Select-Object -ExpandProperty Hash
+            if ($local_hash -ne $upstream_hash) {
+                $anyDifference = $true
+                Copy-Item -Path $upstream_path -destination $local_path -Force
+            }
         }
         #
         # use  GITHUB_OUTPUT to propagate a value that will be able to control the conditional on a future step
@@ -62,13 +75,13 @@ jobs:
             "ANY_DIFFERENCE=True" | Out-File -FilePath $env:GITHUB_OUTPUT -Encoding utf8 -Append
         }
         #
-        # Store the date:
+        # Store the date and head reference for use in subsequent steps
         $dateNow = (Get-Date).ToString("yyyy_MMMM_dd")
         echo "CHANGE DATE will be $dateNow"
         "CHANGE_DATE=$dateNow" | Out-File -FilePath $env:GITHUB_OUTPUT -Encoding utf8 -Append
         "HEAD_REF=$(git --git-dir=npp_repo/.git describe --always '@')" | Out-File -FilePath $env:GITHUB_OUTPUT -Encoding utf8 -Append
         #
-        # get rid of the extra hierarchy now that the hash
+        # get rid of the extra hierarchy now that everything necessary has been copied
         Remove-Item .\npp_repo -Force -Recurse -ErrorAction SilentlyContinue
         #
         git status

